### PR TITLE
Fix COPY --if-exists with wildcard

### PR DESCRIPTION
**Pull request by mikejholly**
Fixes: https://github.com/earthly/earthly/issues/3875

This PR fixes the issue, but I'm curious why the `!strings.Contains(src, "*")` was added in the first place. It seems like all of the relevant tests pass without this extra condition.

### CODE REVIEW

File: tests/Earthfile [hunk 1]
```diff
@@ -256,6 +257,19 @@ copy-test-empty-src:
     DO +RUN_EARTHLY --earthfile=copy.earth --target=+copy-empty-src
     DO +RUN_EARTHLY --earthfile=copy.earth --target=+copy-empty-src-if-exists
 
+copy-test-wildcard:
+    DO +RUN_EARTHLY --earthfile=copy.earth --target=+copy-wildcard
+
+copy-test-if-exists-wildcard:
+    # Regression test for 'COPY --if-exists non_existent_folder/* /'
```
Comment (alexcb) [inline tests/Earthfile:264]: Nit: I would drop the "Regression test" from this comment, maybe just:
```suggestion
    # test 'COPY --if-exists non_existent_folder/* /' works when "non_existent_folder" doesn't exist.
```

File: tests/Earthfile [hunk 2]
```diff
@@ -256,6 +257,19 @@ copy-test-empty-src:
...
+    DO +RUN_EARTHLY --earthfile=copy.earth --target=+copy-if-exists-wildcard
+
+    RUN mkdir non_existent_folder
```
Comment (alexcb) [inline tests/Earthfile:267]: hah.... maybe "non_existent_folder" should be named to something else, since it exists now.

File: tests/copy.earth [hunk 3]
```diff
@@ -222,3 +222,9 @@ copy-empty-src:
 copy-empty-src-if-exists:
     ARG SRC=""
     COPY --if-exists $SRC $SRC
+
+copy-if-exists-wildcard:
+    COPY --if-exists non_existing_folder/* /
```
Comment (alexcb) [inline tests/copy.earth:227]: what about adding:

```
ARG test_foo_exists=false
IF [ "$test_foo_exists" ]
    RUN test -f touch non_existent_folder/foo
END
```

so we can then call:

```
DO +RUN_EARTHLY --earthfile=copy.earth --target="+copy-if-exists-wildcard --test_foo_exists=true"
```

and perhaps another similar arg for testing the expected outcome when `non_existent_folder` exists but is empty?

File: tests/Earthfile (hunk 1 above)
Comment (mikejholly) [reply tests/Earthfile:264]: I removed it as it's somewhat redundant anyhow.

File: util/llbutil/copyop.go [hunk 4]
```diff
@@ -30,17 +30,12 @@ func CopyOp(ctx context.Context, srcState pllb.State, srcs []string, destState p
...
 	for _, src := range srcs {
-		// If the copy came in as optional (ifExists), then we need to trigger
-		// the underlying wildcard matching and allow empty wildcards. The
-		// matching uses the filepath match syntax. So by simply creating a
-		// wildcard where the first letter needs to match the current first
-		// letter gets us the single match; and no error if it is missing. Paths
-		// with wildcards will be automatically ignored when no match is present.
-		if ifExists && len(src) != 0 && !strings.Contains(src, "*") {
-			// Normalize path by dropping './'
+		// HACK: For COPY --if-exists, we can use a glob expression (e.g., '[f]oo') to
```
Comment (alexcb) [inline util/llbutil/copyop.go:33]: I'm confused by the comment -- where are we using a glob expression?

Comment (mikejholly) [reply util/llbutil/copyop.go:33]:
```
src = fmt.Sprintf("[%s]%s", string(src[0]), string(src[1:]))
```
This code uses a bit of a hack to have BK treat the file(s) as optional. `foo` becomes `[f]oo`, which has BK interpret the source as an expression to expand.

File: tests/copy.earth (hunk 3 above)
Comment (mikejholly) [reply tests/copy.earth:227]: Updated with `test -f`. I used separate targets as the parameters were more complicated than the extra 3 lines.

File: tests/Earthfile (hunk 2 above)
Comment (mikejholly) [reply tests/Earthfile]: Updated.

File: util/llbutil/copyop.go (hunk 4 above)
Comment (alexcb) [reply util/llbutil/copyop.go:33]: oh wow.... yeah I didn't understand what that particular line was doing. This is really hacky.

A comment directly above this line would make it easier, maybe something like:

```
// HACK: a regular expression is constructed, e.g. "foo" becomes "[f]oo"
src = fmt.Sprintf("[%s]%s", string(src[0]), string(src[1:]))
```

Comment (mikejholly) [reply util/llbutil/copyop.go]: Sure, I moved the comment.

**Review by alexcb — APPROVED** [review]

File: util/llbutil/copyop.go (hunk 4 above)
Comment (alexcb) [reply util/llbutil/copyop.go]: thanks, it feels easier to follow now.
